src/main/scala/viper/gobra/translator/encodings/StringEncoding.scala

```diff
@@ -0,0 +1,168 @@
+    */
+  override def expr(ctx: Context): in.Expr ==> CodeWriter[vpr.Exp] = {
+
+    def goE(x: in.Expr): CodeWriter[vpr.Exp] = ctx.expr.translate(x)(ctx)
+
+    default(super.expr(ctx)) {
+      case (e: in.DfltVal) :: ctx.String() / Exclusive =>
+        val encodedStr = stringBeginning + "Default"
+        strLengths += encodedStr -> 0
+        unit(withSrc(vpr.DomainFuncApp(func = makeFunc(encodedStr), Seq(), Map.empty), e))
+      case lit: in.StringLit if lit.typ.addressability == Exclusive =>
+        val encodedStr = stringBeginning + Hex.encodeHexString(lit.s.getBytes("UTF-8"))
+        strLengths += encodedStr -> lit.s.length
+        unit(withSrc(vpr.DomainFuncApp(func = makeFunc(encodedStr), Seq(), Map.empty), lit))
+      case len@in.Length(exp :: ctx.String()) =>
+        for { e <- goE(exp) } yield withSrc(vpr.DomainFuncApp(func = lenFunc, Seq(e), Map.empty), len)
+      case concat@in.Concat(l :: ctx.String(), r :: ctx.String()) =>
+        for {
+          lEncoded <- goE(l)
+          rEncoded <- goE(r)
+        } yield withSrc(vpr.DomainFuncApp(concatFunc, Seq(lEncoded, rEncoded), Map.empty),concat)
+    }
+  }
+
+  /**
+    * Encodes a string variable as an vpr.Int
+    */
+  override def variable(ctx: Context): in.BodyVar ==> vpr.LocalVarDecl = {
+    case v :: ctx.String() / Exclusive if typ(ctx).isDefinedAt(v.typ) =>
+      val (pos, info, errT) = v.vprMeta
+      vpr.LocalVarDecl(v.id, vpr.Int)(pos, info, errT)
+    case v :: ctx.String() / Shared if typ(ctx).isDefinedAt(v.typ) =>
+      val (pos, info, errT) = v.vprMeta
+      vpr.LocalVarDecl(v.id, vpr.Ref)(pos, info, errT)
+  }
+
+  override def finalize(col: Collector): Unit = {
+    col.addMember(genDomain())
+  }
+
+  /** Every string literal in the program is encoded as a unique domain function in the String domain,
+    * whose value corresponds to the string id.
+    *     unique function stringLitX(): Int
+    * Here, X is an unique suffix depending on the literal value.
+    */
+  private def makeFunc(name: String): vpr.DomainFunc = {
+    val func = vpr.DomainFunc(
+      name = name,
+      formalArgs = Seq(),
+      typ = vpr.Int,
+      unique = true
+    )(domainName = domainName)
+    funcs += name -> func
+    func
+  }
+
+  /**
+    * Generates
+    *   function strLen(id: Int): Int
+    */
+  private val lenFuncName: String = "strLen"
+  private val lenFunc: vpr.DomainFunc = vpr.DomainFunc(
+    name = lenFuncName,
+    formalArgs = Seq(vpr.LocalVarDecl(Names.freshName, vpr.Int)()),
+    typ = vpr.Int,
+  )(domainName = domainName)
+
+  /**
+    * Generates
+    *   function strConcat(l: Int, r: Int): Int
+    * where l and r are string ids
+    */
+  private val concatFuncName: String = "strConcat"
+  private val concatFunc: vpr.DomainFunc = vpr.DomainFunc(
+    name = concatFuncName,
+    formalArgs = Seq(vpr.LocalVarDecl(Names.freshName, vpr.Int)(), vpr.LocalVarDecl(Names.freshName, vpr.Int)()),
```

Comment:
Here, you do not have to use freshName. You should avoid using freshName if possible.

Reply:
same for many of the other functions. I am not repeating it.

Reply:
My bad. Fixed in the latest commit